Add Journey to find a replacement to the Java Security Manager blog (#3784)
opensearch-project/project-website: _posts/2025-05-19-journey-to-find-replacements-java-security-manager.md

---
layout: post
title: "Finding a replacement for JSM in OpenSearch 3.0"
authors:
   - cwperks
   - kumargu
   - kolchfa
date: 2025-05-19
categories:
  - technical-posts
meta_keywords: Opensearch security, Java security manager, OpenSearch 3.0, security, sandboxing.
meta_description: The Java Security Manager (JSM) has long been a foundational security mechanism in OpenSearch. With its deprecation underway, this blog post explores the alternative technologies adopted to replace JSM.
---

OpenSearch 3.0.0 introduced many innovative features that provide significant advancements in performance, data management, vector database functionality, and more. In [the release announcement](https://opensearch.org/blog/unveiling-opensearch-3-0/), we shared that OpenSearch has replaced the Java Security Manager (JSM) because of its upcoming deprecation. In this blog post, we share more information about the efforts that enabled this transition.

The decision to remove JSM in 3.0 was carefully considered and primarily driven by two factors:

> 1. **The upcoming removal of JSM from the Java platform**: JSM has been deprecated since JDK 17 ([JEP411](https://openjdk.org/jeps/411)) and is scheduled for full removal in JDK 24 ([JEP 486](https://openjdk.org/jeps/486)). According to the Java Enhancement Proposal (JEP), this decision was made because very few projects were using JSM, with Elasticsearch being one of the few major open-source projects that did. While OpenSearch 3.0 is bundled with Java 21 and not yet forced to drop JSM, continuing to rely on a deprecated and soon-to-be-removed component was deemed unsustainable for long-term support and innovation.

Reviewer: ```suggestion
1. **The upcoming removal of JSM from the Java platform**: JSM has been deprecated since JDK 17 ([JEP411](https://openjdk.org/jeps/411)) and is scheduled for full removal in JDK 24 ([JEP 486](https://openjdk.org/jeps/486)). According to the Java Enhancement Proposal (JEP), this decision was made because very few projects were using JSM. While OpenSearch 3.0 is bundled with Java 21 and not yet forced to drop JSM, continuing to rely on a deprecated and soon-to-be-removed component was deemed unsustainable for long-term support and innovation.
```

> 

Author: I think we can skip Elasticsearch. If you could please suggest an edit based on the write way to frame it, i will accept the changes.

2. **Incompatibility with emerging JVM features, particularly virtual threads**: Introduced in ([JEP 444](https://openjdk.org/jeps/444)), virtual threads are one of the most anticipated features in modern Java. While OpenSearch 3.0 does not use virtual threads internally, we expect plugin developers and future versions of OpenSearch to potentially use virtual threads for improved scalability. However, virtual threads do not carry permissions when a Security Manager is enabled, effectively rendering them incompatible with any JSM-based security model. Thus, continuing to support JSM would have prevented adoption of a key Java feature that unlocks better concurrency and resource efficiency.

> Given these needs, we decided to deprecate JSM in OpenSearch 3.0. Doing this in a major version allowed us to communicate the change clearly and avoid introducing breaking security changes in later 3.x minor releases.

Reviewer 2: @pajuric There's a link to the 3.0.0 blog post near the top. Should this be included everywhere or only once?

---
layout: post

> title: "Journey to find a replacement to the Java Security Manager in OpenSearch 3.0"

Reviewer 3: Wonder if `create` or `propose` is better than `find` for the title.

Reviewer 2: I agree. "find" is appropriate here since the blog delves into the lengthy process of identifying a suitable replacement. It also goes into creating the replacement so you could argue to make it `find and create`, but I think it also works with `find` alone.

authors:
   - cwperks
   - kumargu
date: 2025-05-19
categories:
  - technical-posts
meta_keywords: Opensearch security, Java security manager, OpenSearch 3.0, security, sandboxing.
meta_description: The Java Security Manager (JSM) has long been a foundational security mechanism in OpenSearch. However, with its deprecation underway in the JDK, this blog explores the alternative technologies adopted to replace JSM,  with the trade-offs considered maintaining a very high bar for OpenSearch’s security posture.
---

> [OpenSearch 3.0.0](https://opensearch.org/blog/unveiling-opensearch-3-0/) introduced many innovative features that continue to push the frontier with significant advancements in performance, data management, vector database functionality, and more. In the announcement we shared that OpenSearch replaced the Security Manager provided by Java and wanted to share more details about this notable effort.

Reviewer 3: ```suggestion
[OpenSearch 3.0.0](https://opensearch.org/blog/unveiling-opensearch-3-0/) introduced many innovative features that continue to push the frontier with significant advancements in performance, data management, vector database functionality, and more. In the announcement, we shared that OpenSearch has replaced Java’s Security Manager feature due to its upcoming deprecation, and we’d like to share more details about the notable efforts that enable this transition.
```

The decision to remove JSM in 3.0 was carefully considered and primarily driven by two factors:

> The upcoming removal of JSM in the Java platform: The Java Security Manager has been deprecated since JDK 17 ([JEP411](https://openjdk.org/jeps/411)) and is scheduled for full removal in JDK 24 ([JEP 486] (https://openjdk.org/jeps/486)).The JEP cited poor adoption of the functionality for its deprecation and specifically highlighted a few large open-source projects as adopters including Elasticsearch. While OpenSearch 3.0 itself is bundled with Java 21 and not yet forced to drop JSM, continuing to rely on a deprecated and soon-to-be-removed component was deemed unsustainable for long-term support and innovation.
> 
> Incompatibility with emerging JVM features—particularly virtual threads: Introduced in ([JEP 444](https://openjdk.org/jeps/444)), virtual threads are one of the most anticipated features in modern Java. While OpenSearch 3.0 does not use virtual threads internally, we expect plugin developers and future versions of OpenSearch to explore their use for improved scalability. However, virtual threads do not carry permissions when a Security Manager is enabled—effectively rendering them incompatible with any JSM-based security model. That means continuing to support JSM would have prevented adoption of a key Java feature that unlocks better concurrency and resource efficiency.

Reviewer 3: ```suggestion
1. The upcoming removal of JSM in the Java platform: The Java Security Manager has been deprecated since JDK 17 ([JEP411](https://openjdk.org/jeps/411)) and is scheduled for full removal in JDK 24 ([JEP 486] (https://openjdk.org/jeps/486)).The JEP cited poor adoption of the functionality for its deprecation and specifically highlighted a few large open-source projects as adopters including Elasticsearch. While OpenSearch 3.0 itself is bundled with Java 21 and not yet forced to drop JSM, continuing to rely on a deprecated and soon-to-be-removed component was deemed unsustainable for long-term support and innovation.

2. Incompatibility with emerging JVM features—particularly virtual threads: Introduced in ([JEP 444](https://openjdk.org/jeps/444)), virtual threads are one of the most anticipated features in modern Java. While OpenSearch 3.0 does not use virtual threads internally, we expect plugin developers and future versions of OpenSearch to explore their use for improved scalability. However, virtual threads do not carry permissions when a Security Manager is enabled—effectively rendering them incompatible with any JSM-based security model. That means continuing to support JSM would have prevented adoption of a key Java feature that unlocks better concurrency and resource efficiency.
```

> Given these needs, OpenSearch 3.0 was seen as the right time to deprecate the Security Manager. Doing so in a major version allowed us to communicate the change clearly and avoid introducing breaking security changes in later 3.x minor releases.

Reviewer 3: ```suggestion
Given these requirements, OpenSearch 3.0 was seen as the right time to deprecate the Security Manager. Doing so in a major version allowed us to communicate the change clearly and avoid introducing breaking security changes in later 3.x minor releases.
```

> 
> In December 2021, an [issue](https://github.com/opensearch-project/OpenSearch/issues/1687) was created in the OpenSearch core repo to discuss options for a replacement where it soon became clear that there was no direct replacement to the functionality that the Security manager provided for Java programs. OpenSearch embarked on a lengthy search for a replacement that sought to retain important functionality from the Security Manager that OpenSearch relied upon.

Reviewer 3: ```suggestion
Back in December 2021, an [issue](https://github.com/opensearch-project/OpenSearch/issues/1687) was created in the OpenSearch core repo to discuss options for a replacement where it soon became clear that there was no direct replacement to the functionality that the Security manager provided for Java programs. OpenSearch embarked on a lengthy search for a replacement that sought to retain important functionality from the Security Manager that OpenSearch relied upon.
```

> Many different options were considered with notable highlights including:
> 
> 1. A radical change to [OpenSearch extensibility](https://opensearch.org/blog/technical-roadmap-opensearch-extensibility/) through out-of-process extensions. 
> 2. GraalVM
> 3. Systemd hardening
> 4. Java agent
> 5. Remove the Security Manager altogether

Reviewer 3: ```suggestion
Many different options were considered with notable highlights including:

1. A radical change to [OpenSearch extensibility](https://opensearch.org/blog/technical-roadmap-opensearch-extensibility/) through out-of-process extensions.
2. Replace JVM with GraalVM.
3. Security hardening with the use of systemd.
4. Introduce a separate Java agent.
5. Completely remove the Java Security Manager and related functionalities.
```

            Provider provider = (Provider) providerClass.getDeclaredConstructor().newInstance();
            Security.addProvider(provider);
            log.debug("Bouncy Castle Provider added");
        } catch (Exception e) {
            log.debug("Bouncy Castle Provider could not be added", e);
        }
    }
    return null;
});
```

In the block above, there is the usage of `SpecialPermission.check()` before the call to `AccessController.doPrivileged(() -> { ... })`. In OpenSearch, `org.opensearch.SpecialPermission` is granted to all jars across the core and any installed plugin at runtime. If there is an external call on the `callstack` (for example if the code came from an API payload), then the call to `SpecialPermission.check()` fails and will prevent code originating externally to OpenSearch from performing privileged actions. 

2. **Cluster Administrators**: 

For cluster administrators, they are prompted about permissions that a plugin requests at installation time. Below is a snippet from the Security plugin that contains grants that allow it to perform actions like adding the BouncyCastle provider.

> ```
> @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
> @     WARNING: plugin requires additional permissions     @
> @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
> * java.io.FilePermission /proc/sys/net/core/somaxconn#plus read
> * java.lang.RuntimePermission accessDeclaredMembers
> 
> ```

Reviewer 3: ```suggestion
```console
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@     WARNING: plugin requires additional permissions     @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
* java.io.FilePermission /proc/sys/net/core/somaxconn#plus read
* java.lang.RuntimePermission accessDeclaredMembers
```
```
